File: NeetCode/Math_And_Geometry/Rotate_Image/rotate__image.py
from typing import Optional, List


class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix)
        for index in range(n // 2):
            matrix[index], matrix[-(index + 1)] = matrix[-(index + 1)], matrix[index]

        for i in range(n):
            for j in range(i + 1, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]


if __name__ == "__main__":
    sol = Solution()
    matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    sol.rotate(matrix)
    for m in matrix:
        print(m)
    print("Running Solution...")

File: NeetCode/Math_And_Geometry/Rotate_Image/Rotate_Image.cpp
// Rotate_Image.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    void rotate(vector<vector<int>> &matrix)
    {
        int n = matrix.size();

        for (int index = 0; index < n / 2; index++)
        {
            vector<int> tmp = matrix[index];
            matrix[index] = matrix[n - index - 1];
            matrix[n - index - 1] = tmp;
        }

        for (int i = 0; i < n; i++)
        {
            for (int j = i + 1; j < n; j++)
            {
                int k = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = k;
            }
        }
    }
};